x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/IndicesPermission.java

```diff
@@ -240,35 +241,154 @@ public Automaton allowedActionsMatcher(String index) {
+            final List<String> concreteIndices = new ArrayList<>(indices.size());
+            for (var idx : indices) {
+                concreteIndices.add(idx.getIndex().getName());
+            }
+            return new IndexResource(name, abstraction.getType(), concreteIndices, null);
+        }
+
+        /**
+         * @return {@code true} if-and-only-if this object is related to a data-stream, either by having a {@link #type} of
+         * {@link IndexAbstraction.Type#DATA_STREAM} or by being the backing index for a {@link #parentDataStream data-stream}.
+         */
+        public boolean isDataStreamRelated() {
+            switch (type) {
+                case DATA_STREAM:
+                    return true;
+                case CONCRETE_INDEX:
+                    return parentDataStream != null;
+                default:
+                    return false;
+            }
+        }
+
+        /**
+         * Check whether this object is covered by the provided permission {@link Group}.
+         * For indices that are part of a data-stream, this checks both the index name and the parent data-stream name.
+         * In all other cases, it checks the name of this object only.
+         */
+        public boolean checkIndex(Group group) {
+            if (parentDataStream != null) {
+                if (group.checkIndex(parentDataStream)) {
+                    return true;
                 }
-                isBackingIndex = indexAbstraction.getType() == IndexAbstraction.Type.CONCRETE_INDEX &&
-                        indexAbstraction.getParentDataStream() != null;
-                isDataStream = indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM;
+            }
+            return group.checkIndex(name);
+        }
+
+        /**
+         * @return the number of distinct objects to which this expansion refers.
+         */
+        public int size() {
+            if (type == IndexAbstraction.Type.CONCRETE_INDEX) {
+                return 1;
             } else {
-                isBackingIndex = isDataStream = false;
+                return 1 + concreteIndices.size();
             }
+        }
+    }
+
+    /**
+     * Authorizes the provided action against the provided indices, given the current cluster metadata
+     */
+    public Map<String, IndicesAccessControl.IndexAccessControl> authorize(
+        String action,
+        Set<String> requestedIndicesOrAliases,
+        Map<String, IndexAbstraction> lookup,
+        FieldPermissionsCache fieldPermissionsCache
+    ) {
+
+        final List<IndexResource> resources = new ArrayList<>(requestedIndicesOrAliases.size());
+        int totalResourceCount = 0;
+
+        for (String indexOrAlias : requestedIndicesOrAliases) {
+            final IndexResource resource = IndexResource.create(indexOrAlias, lookup.get(indexOrAlias));
+            resources.add(resource);
+            totalResourceCount += resource.size();
```

Comment:
It is possible that concreteIndices of multiple IndexResource overlap. For example: 

requestedIndicesOrAliases = Set.of("alias1", "alias2")
where both aliases point to `index1`.

So the total count of distinct resources is 3 instead of 4.

Reply:
Yes, that is possible. The important thing here is to avoid reallocating the maps. If we over-size them that's a tiny heap cost with essentially no performance cost.